tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h

```diff
@@ -0,0 +1,155 @@
+
+      public:
+        Shape &body_part_shape_;
+
+        TagBufferParticles(ConcurrentIndexVector &buffer_particle_list, RealBody &real_body, SharedPtr<Shape> shape_ptr)
+            : BaseDynamics<void>(real_body), LocalDynamics(real_body), GeneralDataDelegateSimple(real_body),
+              body_part_shape_(shape_ptr_keeper_.assignRef(shape_ptr)), buffer_particle_list_(buffer_particle_list),
+              base_particles_(real_body.getBaseParticles()), 
+              buffer_particle_indicator_(*particles_->registerSharedVariable<int>("BufferParticleIndicator"))
+        {
+            particles_->registerSortableVariable<int>("BufferParticleIndicator");    
+        };
+        virtual ~TagBufferParticles(){};
+    
+        virtual void update(size_t index_i, Real dt = 0.0) 
+        {
+            if (body_part_shape_.checkContain(base_particles_.pos_[index_i]))          
+            {
+                buffer_particle_list_.push_back(index_i);
+                buffer_particle_indicator_[index_i] = 1;               
+            }
+        };
+        
+        virtual void exec(Real dt = 0.0) override
+        {
+            setupDynamics(dt);
+            
+            particle_for(ExecutionPolicy(), base_particles_.total_real_particles_,
+                         [&](size_t index_i)
+                         { update(index_i, dt); });
+        };
+
+      protected:
+        ConcurrentIndexVector &buffer_particle_list_;
+        BaseParticles &base_particles_;
+        StdLargeVec<int> &buffer_particle_indicator_;  
+    };
+
+    template <class ExecutionPolicy>
+    class Injection : public BaseDynamics<void>, public LocalDynamics, public GeneralDataDelegateSimple 
+    {
+      public:
+        Injection(ConcurrentIndexVector &buffer_particle_list, RealBody &real_body, SharedPtr<AlignedBoxShape> shape_ptr,
+                  size_t body_buffer_width, int axis_direction, BidirectionalBuffer& buffer) : 
+            BaseDynamics<void>(real_body), LocalDynamics(real_body),GeneralDataDelegateSimple(real_body),
+            buffer_particle_list_(buffer_particle_list), aligned_box_(*shape_ptr.get()),
+            fluid_(DynamicCast<Fluid>(this, particles_->getBaseMaterial())), 
+            pos_n_(particles_->pos_), rho_n_(particles_->rho_), p_(*particles_->getVariableByName<Real>("Pressure")), 
+            axis_(axis_direction), body_buffer_width_(body_buffer_width),                                                                                                          
+            previous_surface_indicator_(*particles_->getVariableByName<int>("PreviousSurfaceIndicator")),                                                                                                                 
+            buffer_particle_indicator_(*particles_->getVariableByName<int>("BufferParticleIndicator")),
+            buffer_(buffer)           
+        {
+            size_t total_body_buffer_particles = 1000.0 * body_buffer_width_;
```

Comment:
the body_buffer_width_ is the variable which is given by the users themselves in the case file. here, 1000 is a safety factor, in case there are not sufficient spare buffer particles in the memory.